### ISSUE TITLE
Support deserialisation from JS objects

### ISSUE THREAD

**Issue by flaghacker** [Kotlin/kotlinx.serialization#24]
In Kotlin-JS it would be useful to have the ability to deserialize JS objects to Kotlin objects. The two closest things we can get right now are

1. Serialize the JS object to Json and deserialize that using this library, this is not very efficient.
2. Define an `external` Kotlin variable whose data is provided by JS, this doesn't work for instance methods that are only defined in Kotlin.

Our use case is that we have a bunch of JS code that generates/provides some data, it would require quite a bit of effort to convert it to Kotlin.

**Comment by bashor** [comment]
@flaghacker If you provide more information we can find solution/workaround for you.
1) Do you need to verify that these objects correctly implement some "interface"?
2) Do you need to load as Kotlin data class?
3) What for do you need to implement equals, hashCode, toString?

**Comment by flaghacker** [comment]
1. I do need to verify that the objects data correspond to the values required by the class definition:
```
data class MyClass(val a: String, val b: String)
val before = js("{a:"test"};")
val after = JSObj.parse<MyClass>(obj)
```
should fail because `b` is missing from `before`.

2. Yes, I need it to parse to a Kotlin data class, but ideally the solution should work for any Kotlin class.
3. I need the parsed objects to behave exactly like they would if they were actually defined in Kotlin, they'll be passed to common code and used in collections and such.

The underlying problem is that I have data defined in JS and behavior defined in Kotlin, and I would like to combine those.

**Comment by sandwwraith** [comment]
Some simple parser was implemented in version 0.3 ([docs](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/runtime_usage.md#dynamic-object-parser-js-only))

You can play around with it and reopen the issue if something crucial is missing
